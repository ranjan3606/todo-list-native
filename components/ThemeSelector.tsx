import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { useTranslation } from '@/i18n';

type ThemeOption = 'light' | 'dark' | 'system';

interface ThemeSelectorProps {
  currentTheme: ThemeOption;
  colorScheme: 'light' | 'dark';
  onThemeChange: (theme: ThemeOption) => void;
}

export function ThemeSelector({ currentTheme, colorScheme, onThemeChange }: ThemeSelectorProps) {
  const { t } = useTranslation();
  
  return (
    <View style={styles.themeOptions}>
      <TouchableOpacity 
        style={[
          styles.themeOption, 
          currentTheme === 'light' && styles.selectedOption,
          { backgroundColor: Colors[colorScheme].buttonBackground }
        ]}
        onPress={() => onThemeChange('light')}
      >
        <FontAwesome name="sun-o" size={24} color={currentTheme === 'light' ? Colors.primary : Colors[colorScheme].text} />
        <Text style={[styles.optionText, { color: Colors[colorScheme].text }]}>
          {t('settings.theme.light')}
        </Text>
      </TouchableOpacity>
      
      <TouchableOpacity 
        style={[
          styles.themeOption, 
          currentTheme === 'dark' && styles.selectedOption,
          { backgroundColor: Colors[colorScheme].buttonBackground }
        ]}
        onPress={() => onThemeChange('dark')}
      >
        <FontAwesome name="moon-o" size={24} color={currentTheme === 'dark' ? Colors.primary : Colors[colorScheme].text} />
        <Text style={[styles.optionText, { color: Colors[colorScheme].text }]}>
          {t('settings.theme.dark')}
        </Text>
      </TouchableOpacity>
      
      <TouchableOpacity 
        style={[
          styles.themeOption, 
          currentTheme === 'system' && styles.selectedOption,
          { backgroundColor: Colors[colorScheme].buttonBackground }
        ]}
        onPress={() => onThemeChange('system')}
      >
        <FontAwesome name="mobile" size={24} color={currentTheme === 'system' ? Colors.primary : Colors[colorScheme].text} />
        <Text style={[styles.optionText, { color: Colors[colorScheme].text }]}>
          {t('settings.theme.system')}
        </Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  themeOptions: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  themeOption: {
    flex: 1,
    alignItems: 'center',
    padding: 12,
    borderRadius: 8,
    marginHorizontal: 4,
  },
  selectedOption: {
    borderWidth: 2,
    borderColor: Colors.primary,
  },
  optionText: {
    marginTop: 8,
    fontWeight: '500',
  },
});
