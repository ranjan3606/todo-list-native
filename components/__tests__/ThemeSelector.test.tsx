import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { ThemeSelector } from '../ThemeSelector';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

// Mock storage services
jest.mock('@/services/storage', () => ({
  getTags: jest.fn(() => Promise.resolve({})),
  saveTags: jest.fn(() => Promise.resolve())
}));

jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { View } = require('react-native');
  return {
    FontAwesome: function MockFontAwesome(props: any) {
      return React.createElement(View, {...props, testID: `icon-${props.name}`});
    }
  };
});

// Mock Colors with exact structure needed
jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { 
      buttonBackground: '#f0f0f0', 
      text: '#000',
      cardBackground: '#fff',
      lightText: '#777'
    },
    dark: { 
      buttonBackground: '#333', 
      text: '#fff',
      cardBackground: '#222',
      lightText: '#aaa'
    },
    primary: '#007aff',
  }
}));

// Mock translations
jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key: string) => {
      const translations = {
        'settings.theme.light': 'Light',
        'settings.theme.dark': 'Dark',
        'settings.theme.system': 'System',
      };
      return translations[key] || key;
    }
  })
}));

describe('ThemeSelector', () => {
  const mockOnThemeChange = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(
      <ThemeSelector 
        currentTheme="system" 
        colorScheme="light" 
        onThemeChange={mockOnThemeChange} 
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays the system theme option as selected when theme is system', () => {
    const { getByText } = render(
      <ThemeSelector 
        currentTheme="system" 
        colorScheme="light" 
        onThemeChange={mockOnThemeChange} 
      />
    );
    
    // Get the parent TouchableOpacity of the system option
    const systemElement = getByText('System');
    const systemTouchable = systemElement.parent?.parent;
    
    // Check that it has the selectedOption style (has borderColor property)
    expect(systemTouchable?.props.style).toEqual(
      expect.objectContaining({ 
        borderWidth: 2,
        borderColor: '#007aff'
      })
    );
  });

  it('displays the light theme option as selected when theme is light', () => {
    const { getByText } = render(
      <ThemeSelector 
        currentTheme="light" 
        colorScheme="light" 
        onThemeChange={mockOnThemeChange} 
      />
    );
    
    // Get the parent TouchableOpacity of the light option
    const lightElement = getByText('Light');
    const lightTouchable = lightElement.parent?.parent;
    
    // Check that it has the selectedOption style (has borderColor property)
    expect(lightTouchable?.props.style).toEqual(
      expect.objectContaining({ 
        borderWidth: 2,
        borderColor: '#007aff'
      })
    );
  });

  it('displays the dark theme option as selected when theme is dark', () => {
    const { getByText } = render(
      <ThemeSelector 
        currentTheme="dark" 
        colorScheme="light" 
        onThemeChange={mockOnThemeChange} 
      />
    );
    
    // Get the parent TouchableOpacity of the dark option
    const darkElement = getByText('Dark');
    const darkTouchable = darkElement.parent?.parent;
    
    // Check that it has the selectedOption style (has borderColor property)
    expect(darkTouchable?.props.style).toEqual(
      expect.objectContaining({ 
        borderWidth: 2,
        borderColor: '#007aff'
      })
    );
  });

  it('calls onThemeChange when a theme option is clicked', () => {
    const { getByText } = render(
      <ThemeSelector 
        currentTheme="system" 
        colorScheme="light" 
        onThemeChange={mockOnThemeChange} 
      />
    );
    
    // Click the light theme option
    fireEvent.press(getByText('Light').parent?.parent);
    
    // onThemeChange should be called with 'light'
    expect(mockOnThemeChange).toHaveBeenCalledWith('light');
    
    // Click the dark theme option
    fireEvent.press(getByText('Dark').parent?.parent);
    
    // onThemeChange should be called with 'dark'
    expect(mockOnThemeChange).toHaveBeenCalledWith('dark');
    
    // Click the system theme option
    fireEvent.press(getByText('System').parent?.parent);
    
    // onThemeChange should be called with 'system'
    expect(mockOnThemeChange).toHaveBeenCalledWith('system');
  });
});