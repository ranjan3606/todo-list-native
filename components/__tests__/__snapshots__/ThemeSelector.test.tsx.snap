// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSelector renders correctly (snapshot) 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f0f0f0",
        "borderRadius": 8,
        "flex": 1,
        "marginHorizontal": 4,
        "opacity": 1,
        "padding": 12,
      }
    }
  >
    <View
      color="#000"
      name="sun-o"
      size={24}
      testID="icon-sun-o"
    />
    <Text
      style={
        [
          {
            "fontWeight": "500",
            "marginTop": 8,
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      Light
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f0f0f0",
        "borderRadius": 8,
        "flex": 1,
        "marginHorizontal": 4,
        "opacity": 1,
        "padding": 12,
      }
    }
  >
    <View
      color="#000"
      name="moon-o"
      size={24}
      testID="icon-moon-o"
    />
    <Text
      style={
        [
          {
            "fontWeight": "500",
            "marginTop": 8,
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      Dark
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f0f0f0",
        "borderColor": "#007aff",
        "borderRadius": 8,
        "borderWidth": 2,
        "flex": 1,
        "marginHorizontal": 4,
        "opacity": 1,
        "padding": 12,
      }
    }
  >
    <View
      color="#007aff"
      name="mobile"
      size={24}
      testID="icon-mobile"
    />
    <Text
      style={
        [
          {
            "fontWeight": "500",
            "marginTop": 8,
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      System
    </Text>
  </View>
</View>
`;
